### PR TITLE
Port index_select to ATen (CUDA)

### LINKED ISSUE #24578: Migrate `index_select` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @ezyang @gchanan @zou3519 @anjali411 @dylanbespalko

### PR DESCRIPTION
**Pull request by Baranowski**
Fixes #24578

### CODE REVIEW

**Comment by Baranowski** [comment]
This is ready for a review.

ASV benchmark:

```python
import torch

class IndexSelect_1D:
    params = [None]

    def setup(self, n):
        self.src = torch.randn(10**7, device='cuda')
        self.idx = torch.rand(5 * 10**6, device='cuda').mul(10**7).floor().to(torch.long)

    def time_bench(self, n):
        self.src.index_select(0, self.idx)
        torch.cuda.synchronize('cuda')

class IndexSelect_3D:
    params = [0, 1, 2]

    def setup(self, n):
        self.src = torch.randn(500, 500, 500, device='cuda')
        self.idx = torch.rand(250, device='cuda').mul(500).floor().to(torch.long)

    def time_bench(self, n):
        self.src.index_select(n, self.idx)
        torch.cuda.synchronize('cuda')

class IndexSelect_SmallIndices:
    params = [0, 2, 4, 6]

    def setup(self, n):
        self.src = torch.randn(10, 10, 10, 10, 10, 10, 10, 10, 10, device='cuda')
        self.idx = torch.tensor([7, 3, 5, 9, 0, 1, 8, 3, 4], device='cuda')

    def time_bench(self, n):
        self.src.index_select(n, self.idx)
        torch.cuda.synchronize('cuda')
```

`master` results:

```
[ 66.67%] ··· index_select.IndexSelect_1D.time_bench                                                   ok
[ 66.67%] ··· ======== ==========
               param1
              -------- ----------
                None    1.33±0ms
              ======== ==========

[ 83.33%] ··· index_select.IndexSelect_3D.time_bench                                                   ok
[ 83.33%] ··· ======== =============
               param1
              -------- -------------
                 0      1.39±0.01ms
                 1      1.74±0.04ms
                 2      3.22±0.05ms
              ======== =============

[100.00%] ··· index_select.IndexSelect_SmallIndices.time_bench                                         ok
[100.00%] ··· ======== =============
               param1
              -------- -------------
                 0       20.1±0.7ms
                 2      19.7±0.03ms
                 4      24.4±0.03ms
                 6      34.3±0.03ms
              ======== =============
```

This PR results:

```
[ 66.67%] ··· index_select.IndexSelect_1D.time_bench                                                   ok
[ 66.67%] ··· ======== =============
               param1
              -------- -------------
                None    1.33±0.01ms
              ======== =============

[ 83.33%] ··· index_select.IndexSelect_3D.time_bench                                                   ok
[ 83.33%] ··· ======== =============
               param1
              -------- -------------
                 0      1.44±0.04ms
                 1      1.73±0.03ms
                 2      3.24±0.05ms
              ======== =============

[100.00%] ··· index_select.IndexSelect_SmallIndices.time_bench                                         ok
[100.00%] ··· ======== =============
               param1
              -------- -------------
                 0       20.2±0.5ms
                 2      19.8±0.01ms
                 4      24.3±0.05ms
                 6      34.3±0.03ms
              ======== =============
```

**Comment by anjali411** [comment]
ping @VitalyFedyunin for review

**Comment by Baranowski** [comment]
I don't think the CI failures are related.

@anjali411 and @VitalyFedyunin FYI I won't have the time to work on this after two weeks from now, so I will most likely abandon this PR if it doesn't get a review soon.

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (peterbell10) [inline aten/src/ATen/native/cuda/Indexing.cu:725]: Nit: These `TensorInfo`s would be a good place to use `auto`.

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (peterbell10) [inline aten/src/ATen/native/cuda/Indexing.cu:795]: Do all the tensors need to be on the current GPU?
```suggestion
  TORCH_CHECK(at::cuda::check_device({out, self, input}), "Input and output must be on the "
```

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (peterbell10) [inline aten/src/ATen/native/cuda/Indexing.cu:801]: This comes from `MAX_TENSORINFO_DIMS`, right? Maybe use that macro constant directly instead of the magic number 25.

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (Baranowski) [reply aten/src/ATen/native/cuda/Indexing.cu:795]: I'm not sure, I think they just need to be on the same device. There is no variable `input`.

**Comment by Baranowski** [comment]
Rebased and addressed @peterbell10's comments. I don't see why the windows builds are failing. Perhaps they are flaky.

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Indexing.cu:795]: >I'm not sure, I think they just need to be on the same device. 

Well you're replacing `THCTensor_(checkGPU)` which does require current GPU.
https://github.com/pytorch/pytorch/blob/f9eb8824f1e1cd0d73275afaeb5c81fc4b3f0de1/aten/src/THC/generic/THCTensor.cpp#L446

>There is no variable `input`.

Typo, I meant `index`.

Comment (Baranowski) [reply aten/src/ATen/native/cuda/Indexing.cu:795]: Thanks, I didn't know that

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (peterbell10) [inline aten/src/ATen/native/cuda/Indexing.cu:817]:
```suggestion
Tensor index_select_cuda(const Tensor& self, int64_t dim, const Tensor& index) {
```

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (peterbell10) [inline aten/src/ATen/native/cuda/Indexing.cu:789]:
```suggestion
Tensor& index_select_out_cuda(Tensor& out, const Tensor& self, int64_t dim,
```

**Review by peterbell10 — CHANGES_REQUESTED** [review]
Build failure on windows is genuine. `int64_t` is not `long` on windows.

**Review by peterbell10 — APPROVED** [review]

**Comment by Baranowski** [comment]
@VitalyFedyunin as a gentle reminder, this PR is still waiting for a review from FB and I won't be able to work on this after this week.

**Comment by ezyang** [comment]
@mruberry do you think you would be able to help unblock this? (LMK if you can't)

**Comment by mruberry** [comment]
> @mruberry do you think you would be able to help unblock this? (LMK if you can't)

Let me take a look and I'll let you know very soon.

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/Indexing.cu:578]: Can you elaborate on this TODO?

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/Indexing.cu:593]: Seems like we can get rid of this comment?

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/Indexing.cu:622]: Same ask here: what's up with this TODO?

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/Indexing.cu:648]: Can we ditch this comment?

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/Indexing.cu:676]: Seems like this should use numel, since it's checking for whether the tensor is empty and tensors with non-zero dims can be empty (e.g. a tensor of shape [3, 0, 1]).

Disallowing this operation on empty tensors seems OK for now but in a future PR we may want to allow them. That is, a tensor with dims [3, 0, 1] can be validly selected on its left most dimension with [0, 2] to become a tensor with shape [2, 0, 1].

Comment (mruberry) [reply aten/src/ATen/native/cuda/Indexing.cu:676]: My mistake, the current `index_select` works fine on tensors with no elements:

```
t = torch.randn((3, 0, 3))
t.index_select(0, torch.tensor([1, 1, 1, 1, 1]))
: tensor([], size=(5, 0, 3))
```

Is this check still relevant?

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/Indexing.cu:798]: Nit: since the code resizes out later maybe this isn't necessary?

File: aten/src/ATen/cuda/detail/IndexUtils.cuh
Comment (mruberry) [inline aten/src/ATen/cuda/detail/IndexUtils.cuh:16]: Please add a comment describing what the positiveDim bool does

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:12678]: Awesome!

File: aten/src/ATen/cuda/detail/IndexUtils.cuh
Comment (mruberry) [reply aten/src/ATen/cuda/detail/IndexUtils.cuh:16]: Having had a chance to read the entire PR, this is the one change I'm still curious about. Can you describe your thinking here? I believe I understand what's going on but would like to verify.

**Comment by mruberry** [comment]
This looks like an extremely competent port. Really cool that index_select will work for complex types on CUDA, too! (cc @anjali411) 

I have a few questions, mostly about comments and the one new arg to `getTensorInfo`. Looking forward to hearing your thoughts, @Baranowski. I should get you a response very quickly and we'll get this in!

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (Baranowski) [reply aten/src/ATen/native/cuda/Indexing.cu:578]: Sorry, it was a note-to-self that's no longer relevant. I forgot to remove it.

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (Baranowski) [reply aten/src/ATen/native/cuda/Indexing.cu:676]: Yes, this selfDims check can be dropped, good catch.

File: aten/src/ATen/cuda/detail/IndexUtils.cuh
Comment (Baranowski) [reply aten/src/ATen/cuda/detail/IndexUtils.cuh:19]: I have added a comment:

```
// Argument `positiveDim` turns on the legacy behavior of THC's getTensorInfo:
// if `t` is a scalar, it's interpreted as a single-element vector.
// (See the legacy implementation in aten/src/THC/THCTensorTypeUtils.cuh)
```

Unfortunately I didn't note down the details, but I remember that some tests were failing without handling this special case.

Comment (mruberry) [reply aten/src/ATen/cuda/detail/IndexUtils.cuh:19]: The special case this is handling is when the tensor is a zero-dim "scalar tensor," right?

Comment (Baranowski) [reply aten/src/ATen/cuda/detail/IndexUtils.cuh:19]: yes, correct

Comment (mruberry) [reply aten/src/ATen/cuda/detail/IndexUtils.cuh:19]: Got it. Instead of extending `getTensorInfo`, what would you think about checking for a 0-dim tensor and just calling `torch.full` to handle that case? Is that too involved a change at this point?

Comment (Baranowski) [reply aten/src/ATen/cuda/detail/IndexUtils.cuh]: I don't think it's too involved but TBH I don't understand how you would like me to use torch.full.

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (Baranowski) [inline aten/src/ATen/native/cuda/Indexing.cu:674]: As an alternative I have introduced this function to be wrapped around getTensorInfo. Please let me know if you don't think that's a good approach and I will be happy to revert.

Comment (mruberry) [reply aten/src/ATen/native/cuda/Indexing.cu:674]: Sounds like a great solution.

**Review by mruberry — APPROVED** [review]
Awesome!

**Pull request closed without merging**
